sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java

```diff
@@ -469,55 +466,100 @@ public void populateDisplayData(DisplayData.Builder builder) {
-      // But, as each shard (replica or primary) is responsible for only one part of the data,
-      // there will be no duplicate.
-
-      JsonNode statsJson = getStats(true);
-      JsonNode shardsJson =
-          statsJson
-              .path("indices")
-              .path(spec.getConnectionConfiguration().getIndex())
-              .path("shards");
-
-      Iterator<Map.Entry<String, JsonNode>> shards = shardsJson.fields();
-      while (shards.hasNext()) {
-        Map.Entry<String, JsonNode> shardJson = shards.next();
-        String shardId = shardJson.getKey();
-        sources.add(new BoundedElasticsearchSource(spec, shardId));
+      if (backendVersion == 2){
+        // 1. We split per shard :
+        // unfortunately, Elasticsearch 2. x doesn 't provide a way to do parallel reads on a single
+        // shard.So we do not use desiredBundleSize because we cannot split shards.
+        // With the slice API in ES 5.0 we will be able to use desiredBundleSize.
+        // Basically we will just ask the slice API to return data
+        // in nbBundles = estimatedSize / desiredBundleSize chuncks.
+        // So each beam source will read around desiredBundleSize volume of data.
+
+        ConnectionConfiguration connectionConfiguration = spec.getConnectionConfiguration();
+        JsonNode statsJson = BoundedElasticsearchSource.getStats(connectionConfiguration, true);
+        JsonNode shardsJson =
+            statsJson
+                .path("indices")
+                .path(connectionConfiguration.getIndex())
+                .path("shards");
+
+        Iterator<Map.Entry<String, JsonNode>> shards = shardsJson.fields();
+        while (shards.hasNext()) {
+          Map.Entry<String, JsonNode> shardJson = shards.next();
+          String shardId = shardJson.getKey();
+          sources.add(new BoundedElasticsearchSource(spec, shardId, null, null, backendVersion));
+        }
+        checkArgument(!sources.isEmpty(), "No shard found");
+      } else if (backendVersion == 5){
+        long indexSize = BoundedElasticsearchSource
+            .estimateIndexSize(spec.getConnectionConfiguration());
+        float nbBundlesFloat = (float) indexSize / desiredBundleSizeBytes;
+        int nbBundles = (int) Math.ceil(nbBundlesFloat);
+        //ES slice api imposes that the number of slices is <= 1024 even if it can be overloaded
+        if (nbBundles > 1024) {
+          nbBundles = 1024;
+        }
+        // split the index into nbBundles chunks of desiredBundleSizeBytes by creating
+        // nbBundles sources each reading a slice of the index
+        // (see https://goo.gl/MhtSWz)
+        // the slice API allows to split the ES shards
+        // to have bundles closer to desiredBundleSizeBytes
+        for (int i = 0; i < nbBundles; i++) {
+          sources.add(new BoundedElasticsearchSource(spec, null, nbBundles, i, backendVersion));
+        }
```

Comment:
else? (other backend versions)

Reply:
You mean adding an else after `else if (backendVersion == 5){}`?
Only ES v2.x and ES v5.x are supported, ES v1.x isn't.
Please note that there is `checkArgument((backendVersion == 2 || backendVersion == 5)` in `getBackendVersion()`

Reply:
Oh ok. Maybe it makes sense to add an enum and do this using a switch. But maybe ok to keep as-is.